src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -99,6 +150,124 @@ public void OnDiagnosticsUnhandledException(HttpContext httpContext, Exception e
+
+            StringValues requestId;
+            StringValues standardRootId;
+            if (httpContext.Request.Headers.TryGetValue(RequestResponseHeaders.RequestIdHeader, out requestId))
+            {
+                activity.SetParentId(requestId);
+                requestTelemetry.Context.Operation.ParentId = requestId;
+
+                string[] baggage = httpContext.Request.Headers.GetCommaSeparatedValues(RequestResponseHeaders.CorrelationContextHeader);
+                if (baggage != StringValues.Empty)
+                {
+                    foreach (var item in baggage)
+                    {
+                        NameValueHeaderValue baggageItem;
+                        if (NameValueHeaderValue.TryParse(item, out baggageItem))
+                        {
+                            activity.AddBaggage(baggageItem.Name, baggageItem.Value);
+                            if (!requestTelemetry.Context.Properties.ContainsKey(baggageItem.Name))
+                            {
+                                requestTelemetry.Context.Properties[baggageItem.Name] = baggageItem.Value;
+                            }
+                        }
+                    }
+                }
+            }
+            else if (httpContext.Request.Headers.TryGetValue(RequestResponseHeaders.StandardRootIdHeader, out standardRootId))
+            {
+                activity.SetParentId(standardRootId);
+
+                StringValues standardParentId;
+                if (httpContext.Request.Headers.TryGetValue(RequestResponseHeaders.StandardParentIdHeader, out standardParentId))
+                {
+                    requestTelemetry.Context.Operation.ParentId = standardParentId;
+                }
+            }
+            activity.Start();
+            httpContext.Features.Set(activity);
+
+            requestTelemetry.Id = activity.Id;
+            requestTelemetry.Context.Operation.Id = activity.RootId;
+
+            this.client.Initialize(requestTelemetry);
+            requestTelemetry.Start(timestamp);
+            httpContext.Features.Set(requestTelemetry);
+
+            return requestTelemetry;
+        }
+
+        private void SetAppIdInResponseHeader(HttpContext httpContext, RequestTelemetry requestTelemetry)
+        {
+            IHeaderDictionary responseHeaders = httpContext.Response?.Headers;
+            if (responseHeaders != null &&
+                !string.IsNullOrEmpty(requestTelemetry.Context.InstrumentationKey) &&
+                (!responseHeaders.ContainsKey(RequestResponseHeaders.RequestContextHeader) || HttpHeadersUtilities.ContainsRequestContextKeyValue(responseHeaders, RequestResponseHeaders.RequestContextTargetKey)))
+            {
+                string correlationId = null;
+                if (this.correlationIdLookupHelper.TryGetXComponentCorrelationId(requestTelemetry.Context.InstrumentationKey, out correlationId))
+                {
```

Comment:
OperationCorrelationTelemetryInitializer set Request.Source, here it sets Request.Reponse.Header. But basically, they are doing similar things. I guess this part of code can be moved into OperationCorrelationTelemetryInitializer, and I'm willing to remove the additional "InitializeRequestTelemetryOnBeginRequest" in this case.
However, the logic here will override the value, not like other initializers which first checks if the value is null or empty. So I don't want to take the risk of moving it to the initializer.